File: OOP/Figures.m
classdef Figures

properties
    dt
end

methods

    function f = Figures(str)
        f.dt = str;
    end

    function x = lt(figa,figb)
        x = (area(figa)<area(figb));
    end

    function x = plus(figa,figb)
        x = area(figa)+area(figb);
    end

end

methods (Abstract)

    x = area(a);

end

end

File: OOP/figures.jl
module figures

export Figure
abstract Figure

export Circle
type Circle <: Figure
    r
end

export Rectangle
type Rectangle <: Figure
    a
    b
end

export area
area(c::Circle) = pi*c.r^2
area(r::Rectangle) = r.a*r.b

import Base.<, Base.+
<(a::Figure,b::Figure) = (area(a)<area(b))
+(a::Figure,b::Figure) = area(a)+area(b)

end
